src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/RazorFormattingEndpoint.cs

```diff
@@ -147,6 +184,120 @@ public async Task<TextEditContainer> Handle(DocumentRangeFormattingParams reques
+                return null;
+            }
+
+            if (!s_allTriggerCharacters.Contains(request.Character, StringComparer.Ordinal))
+            {
+                _logger.LogWarning($"Unexpected trigger character '{request.Character}'.");
+                return null;
+            }
+
+            var documentSnapshot = await _projectSnapshotManagerDispatcher.RunOnDispatcherThreadAsync(() =>
+            {
+                _documentResolver.TryResolveDocument(request.TextDocument.Uri.GetAbsoluteOrUNCPath(), out var documentSnapshot);
+
+                return documentSnapshot;
+            }, cancellationToken).ConfigureAwait(false);
+
+            if (documentSnapshot is null)
+            {
+                _logger.LogWarning($"Failed to find document {request.TextDocument.Uri}.");
+                return null;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync();
+            if (codeDocument.IsUnsupported())
+            {
+                _logger.LogWarning($"Failed to retrieve generated output for document {request.TextDocument.Uri}.");
+                return null;
+            }
+
+            var sourceText = await documentSnapshot.GetTextAsync().ConfigureAwait(false);
+            var hostDocumentIndex = request.Position.GetAbsoluteIndex(sourceText);
+            var triggerCharacterKind = _razorDocumentMappingService.GetLanguageKind(codeDocument, hostDocumentIndex);
+            if (triggerCharacterKind is not RazorLanguageKind.CSharp)
+            {
+                _logger.LogInformation($"Unsupported trigger character language {triggerCharacterKind:G}.");
+                return null;
+            }
+
+            if (!IsApplicableTriggerCharacter(request.Character, triggerCharacterKind))
+            {
+                // We were triggered but the trigger character doesn't make sense for the current cursor position. Bail.
+                _logger.LogInformation("Unsupported trigger character location.");
+                return null;
+            }
+
+            if (!_razorDocumentMappingService.TryMapToProjectedDocumentPosition(codeDocument, hostDocumentIndex, out _, out var projectedIndex))
+            {
+                _logger.LogWarning($"Failed to map to projected position for document {request.TextDocument.Uri}.");
+                return null;
+            }
+
+            var formattingContext = FormattingContext.Create(
+                request.TextDocument.Uri, documentSnapshot, codeDocument, request.Options, _adhocWorkspaceFactory, isFormatOnType: true);
+            var documentOptions = await GetDocumentOptionsAsync(request, formattingContext.CSharpWorkspaceDocument).ConfigureAwait(false);
+
+            // Ask C# for formatting changes.
+            var formattingChanges = await RazorCSharpFormattingInteractionService.GetFormattingChangesAsync(
```

Comment:
If you're in the mood, this PR could go further and move a lot of this code into `CSharpOnTypeFormattingPass` now. In theory this method shouldn't have language specific bits (in case we add a `HtmlOnTypeFormattingPass` in future.

But could be something done later too.. the fact that all of this moved, and all of the tests pass, is definitely a big win in itself!

Reply:
Ooooh, good point. In the interest of trying to get more bug fixes in before the upcoming P5 deadline, I feel like it might be appropriate to push off the work to 17.1. I do think it would be super valuable to do this though, filed https://github.com/dotnet/aspnetcore/issues/36947.